# Functions

Functions are the core **abstraction tool** used in procedural programming.  When we `call` a function we give in `arguments` which will be processed inside the function and the result is then `returned` to the statement that made the call.  The code that processes the argument are hidden from view in the function and are expected to work as documented in the `doc string`.  

In writing **clean code** functions are designed to do one thing well.  This concept is called **separation of concerns**. Designing functions in your programming is also called **modular programming** and the functions once written are often grouped together into a larger **category of concerns** in Python `modules`.

>**Note:** Modular programming allows the programmer to use both **top-down design**, i.e., from the high level calling of functions to the nature of specific functions, and **bottom-up design**, building functions that address specific concerns and then deciding how to call them to get something done.    

Below is an example of a function.  Run the cell below, it records the name of the function, `expenditure`, and checks to see that the statements making up the function are syntactically correct, but it does not yet do anything. Ypu have to call the function to run it.    

In [1]:
def expenditure(p_1, p_2, q_1, q_2):
    """ Calculates amount_spent = p_1*q_1 + p_2*q_2
    
    args: p_1, float, price of good 1
          p_2, float, price of good 2
          q_1, int, quantity of good 1
          q_2, int, quantity of good 2
          
    returns: amount_spent, float
    
    """

    amount_spent = p_1*q_1 + p_2*q_2

    return amount_spent

Once a function is defined it can be called with values for the arguments it expects.  It will then run the code in the function and return a result if a return statement is used.

In [2]:
expenditure(0.5, 1.25, 10, 2)

7.5

## Syntax of a function

We will now look line by line at our example expenditure function.

> **LIne 1:** Functions begin with the 'def' keyword followed by a function name, a list of arguments in parentheses, and a colon.  Note, function names follow the same rule as variable names.  They should be meaningful, lower case, with words separated by an underscore `_`.<br><br>The list of arguments is a comma seperated list of variable names in parentheses, i.e., `p_1, p_2, q_1, q_2` are called the arguments of the function.  Arguments are the means of passing data to the function when the function is called.  

Notice that lines 2-16 are indented four spaces.  We will refer to all the Python statements at the same level of indentation as a **CODE BLOCK**.  This is the way Python knows which statements following the `def` statement belong to the function.
 
>**Lines 2-12:** make up the **doc-string** for the function.  This allows us to tell someone who wants to use our function what it does.  Notice doc-strings always starts with `"""` folowed by a short description of the function.  The `doc_string` ends with another `"""` on a separate line.  Python uses doc-strings to document a function when the help() function is called.  This is shown below.<br><br> Notice, after the short description we have a description of the arguments the function expects and a description of the return value of the function.

After the doc-string the rest of the code in the CODE BLOCK is performed.
>**Line 14:** actually calculates amount_spent from the argument, and, <br><br> **Line 16:** returns a pointer to the value amount_spent.


When we use a function the variables in the function are called **local variables**.  They only exist inside the function and disappear when we exit the function.  They cannot be accessed outside the function.  To see this try the following example.

In [10]:
balance = 50
print(f"Stating balance = {balance}")
balance = balance - expenditure(0.5, 1.25, 10, 2)
print(f"Stating balance = {balance}")
print(f"You spent = {amount_spent} dollars.")

Stating balance = 50
Stating balance = 42.5


NameError: name 'amount_spent' is not defined

What you get is a `NameError` which we will explain in more detail later in this lesson.  To see how doc-strings are used run the help function in the cell below.

In [3]:
help(expenditure)

Help on function expenditure in module __main__:

expenditure(p_1, p_2, q_1, q_2)
    Calculates amount_spent = p_1*q_1 + p_2*q_2
    
    args: p_1, float, price of good 1
          p_2, float, price of good 2
          q_1, int, quantity of good 1
          q_2, int, quantity of good 2
          
    returns: amount_spent, float



# Why use docstrings?

As already mentioned **doc strings** help document how a function should be used.  It tells the user what the function does, what arguments it accepts, and what result it will return.  You can use the `help()` function to print out a functions docstring as shown in the example below.  The more thought you put into your docstrings the more useful they will be to someone else.

> Beginning programmers often say, or think, <br><br>"but I'm only programming for myself and documentation takes time and effort that I would rather spend on programming".<br><br>  Don't be fooled by this thought.  **You will be that someone else in one day!**

# Quadratic Equations As Functions

In [4]:
def solve_quadratic(a, b, c):
    """
    Given the quadratic equation y = a*x**2 + b*x + c,
    return the values of x, root_1 and root_2, that
    evaluate the quadratic to y = 0.
    
    Args:
        a: float, coefficient a*x**2.
        b: float, coefficient b*x.
        c: float, coefficeint a.

    """
    
    fixed_numerator = (b**2.0 - 4.0 * a * c)**0.5
    fixed_denominator = 2.0*a
    root_1 = (-b + fixed_numerator)/fixed_denominator
    root_2 = (-b - fixed_numerator)/fixed_denominator
    
    return root_1, root_2


In [5]:
# Calling the function
solve_quadratic(1.0, 3.0, -4.0)

(1.0, -4.0)

We can also write a function that will return y given x.

In [6]:
def evaluate_quadratic(a, b, c, x):
    
    """Given the quadratic equation y = a*x**2 + b*x + c,
       and a value for x, return the vlaue y
    """
    
    y = a*x**2.0 + b*x + c
    
    return y

# This is our test case.  You should get 0.0 as a result
a = 1.0
b = 3.0
c = -4.0
root_1, root_2 = solve_quadratic(a, b, c)
value_1 = evaluate_quadratic(a, b, c, root_1)
value_2 = evaluate_quadratic(a, b, c, root_1)
print(f"At x = {root_1}, {a}x^2 +{b}x + {c} = {value_1}")
print(f"At x = {root_2}, {a}x^2 +{b}x + {c} = {value_2}")

At x = 1.0, 1.0x^2 +3.0x + -4.0 = 0.0
At x = -4.0, 1.0x^2 +3.0x + -4.0 = 0.0


Now lets add a function to input the quadratic coefficients.

In [15]:
def input_quadratic_coefficients():
    
    """Returns user input for the values of a, b, c 
    to the quadratic equation ax^2 + bx + c
    """
    
    print("Please enter coefficients for quadratic equation")
    print(" given by ax^2 + bx + c")
    print()
    a = float(input("Please enter a: "))
    b = float(input("Please enter b: "))
    c = float(input("Please enter c: "))
    
    return a, b, c

# Test code
a, b, c = input_quadratic_coefficients()

print()
print(f"You entered the equation y = {a}x^2 +{b}x + {c}")

Please enter coefficients for quadratic equation
 given by ax^2 + bx + c

Please enter a: 3
Please enter b: 2
Please enter c: -15

You entered the equation y = 3.0x^2 +2.0x + -15.0


Now we can write some code to use our three functions.  Going back to our exercise in 101 we can evaluate a = 3.0, b = 4.0, c = 2.0

In [16]:
a, b, c = input_quadratic_coefficients()
print()
print(f"You entered the equation y = {a}x^2 +{b}x + {c}")
root_1, root_2 = solve_quadratic(a, b, c)
print(f"the roots of this equation are {root_1} and {root_2}")
value_1 = evaluate_quadratic(a, b, c, root_1)
value_2 = evaluate_quadratic(a, b, c, root_1)
print()
print("Now we can check to make sure this is correct.")
print(f"{a}x^2 +{b}x + {c} = {value_1} at x = {root_1}")
print(f"{a}x^2 +{b}x + {c} = {value_2} at x = {root_2}")


Please enter coefficients for quadratic equation
 given by ax^2 + bx + c

Please enter a: 2
Please enter b: 1
Please enter c: -6

You entered the equation y = 2.0x^2 +1.0x + -6.0
the roots of this equation are 1.5 and -2.0

Now we can check to make sure this is correct.
2.0x^2 +1.0x + -6.0 = 0.0 at x = 1.5
2.0x^2 +1.0x + -6.0 = 0.0 at x = -2.0


# Namespaces and Scope

In [24]:
help(solve_quadratic)

Help on function solve_quadratic in module __main__:

solve_quadratic(a, b, c)
    Given the quadratic equation y = a*x**2 + b*x + c,
    return the values of x, root_1 and root_2, that
    evaluate the quadratic to y = 0.
    
    Args:
        a: float, coefficient a*x**2.
        b: float, coefficient b*x.
        c: float, coefficeint a.



A very important source of documentation for the Python language itself are called the [Python Enhamcement Proposals](https://www.python.org/dev/peps/).  Over time we will look at a number of these proposals. To get started 

1. [pep-1](https://www.python.org/dev/peps/pep-0001/) provides an introduction to peps and their management. 
2. [pep-8](https://www.python.org/dev/peps/pep-0008/) is the complete (more or less) guide to programming style in Python.
3. [pep-20](https://www.python.org/dev/peps/pep-0020/) is a short poem on Python style and now has an easter egg when you run `import this` in Python.
4. [pep-257](https://www.python.org/dev/peps/pep-0257/) covers doc-string conventions which you should take a quick look at now.

In [25]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


# Python Modules

Now that we have written some functions to work with quadratic equations we would like to save them as a Python file so we can use them at a later date.  

## Building a Module
We can save the functions above into a file which we will call quadratic.py.  We can do this in our notebook in the cell below.  If you notice I added the cell magic command, `%%writefile "quadratic.py"`
.  When the cell is executed the cell magic command  `%%writefile` will write the contents of the cell as a python file called `quadratic.py`.  

In [10]:
%%writefile "quadratic.py

def run():
    """Allows you to input coefficents to a quadratic equation,  
       solve for the roots of the quadratic eqaution, and
       evaluate the quadratic equation at its calculated roots.
    """
    a, b, c = input_coefficients()
    print()
    print(f"You entered the equation y = {a}x^2 +{b}x + {c}")
    root_1, root_2 = find_roots(a, b, c)
    print(f"the roots of this equation are {root_1} and {root_2}")
    value_1 = evaluate(a, b, c, root_1)
    value_2 = evaluate(a, b, c, root_1)
    print()
    print("Now we can check to make sure this is correct.")
    print(f"{a}x^2 +{b}x + {c} = {value_1} at x = {root_1}")
    print(f"{a}x^2 +{b}x + {c} = {value_2} at x = {root_2}")


def find_roots(a, b, c):
    
    """Given the quadratic a*x**2 + b*x + c, return the values
       of x that evaluate the quadratic to 0."""
    
    fixed_numerator = (b**2.0 - 4.0 * a * c)**0.5
    fixed_denominator = 2.0*a
    root_1 = (-b + fixed_numerator)/fixed_denominator
    root_2 = (-b - fixed_numerator)/fixed_denominator
    return root_1, root_2

def evaluate(a, b, c, x):
    
    """Given the quadratic equation y = a*x**2 + b*x + c at x,
       evaluate the equation at x and return y."""
    
    y = a*x**2.0 + b*x + c
    
    return y

def input_coefficients():
    
    """Returns usere input for a, b, c to 
    determine a quadratic equation y = ax^2 + bx + c
    """
    
    print("Please enter coefficients for quadratic equation")
    print(" given by ax^2 + bx + c")
    print()
    a = float(input("Please enter a: "))
    b = float(input("Please enter b: "))
    c = float(input("Please enter c: "))
    
    return a, b, c


Overwriting quadratic.py


## Using a Module

We can now use the `import` statement to import our equations back into our notebook and use them in our code. Once we import a module we can use the builtin `help` function to learn more about the module. This is where the `doc strings` are used.  

In [11]:
import quadratic
help(quadratic)

Help on module quadratic:

NAME
    quadratic

FUNCTIONS
    evaluate(a, b, c, x)
        Evaluates the quadratic expression y = a*x**2 + b*x + c at x, and 
        return y.
    
    find_roots(a, b, c)
        Given the quadratic a*x**2 + b*x + c, return the values
        of x that evaluate the quadratic to 0.
    
    input_coefficients()
        Returns usere input for a, b, c to 
        the quadratic equation ax^2 + bx + c
    
    run()
        Allows you to input coefficents to a quadratic equation,  
        solve for the roots of the quadratic eqaution, and
        evaluate the quadratic equation at its calculated roots.

FILE
    c:\users\desktop\dropbox\2021 mccabe work\notebooks\quadratic.py




# Economics Review

## Mathematical Review

In mathematics, a **function** is a binary relation between two sets that associates to each element of the first set exactly one element of the second set.  The words map, mapping, transformation, correspondence, and operator are all synonymous with the word function.  Let X and Y be sets, then f: X $\to$ Y as a function, if and only if, for all x $\in$ X, f(x) $\in$ Y. We call X the **domain** of f and Y the **codomain** of f.

>**Note:** A function is uniquely represented by the set of all pairs $G=(x, f (x))$, called the **graph of the function**. When the domain and the codomain are sets of real numbers, each such pair may be thought of as the Cartesian coordinates of a point in the plane.  A popular means of illustrating a function is by showing its graph. 

Often a function is defined by an equation, i.e., $f:X \to Y$ defined by $f(x) = x^2$ results in the pairs $(x, x^2) \in X \mathsf X Y$.  

## Utility Functions

A utility function represents a preference relation on tuples representing quantities of different goods.  For example, if we have two goods, food and water, where $x_f$ is the quantity of food consumed and $x_w$ is the quantity of water consumed then the **consumption bundle** is $c = (x_f, x_w) \in \mathbb R_+^2$. If given a choice between two consumption bundles c and c' a consumer chooses c' over c we say c' is **revealed preferred** to c.  Alternatively we can define a utility function for our consumer as $u: \mathbb R_+^2 \to \mathbb R_+$ such that $u(c') > u(c)$, if and only if, our consumer **prefers** c' to c.   

>**Note:** Utility function are not observed and yet economists assume that people act to `maximize utility`.  This is also known as `goal directed behavior` in the behavioral ecology literature.  Furthermore by assuming that utility functions have certain properties economists try to recover a persons the utility function underlying observed choices. 

### Example Utility Equation
Imagine a world with three goods.  The quantities of these goods are represented by the real valued 3-tuple, $x = (x_1, x_2, x_3) \in {\mathbb R_+^3}$, where $\mathbb R_+$ denotes the non negative real numbers.  

>Utility
>1. $u:$$\mathbb R_+^3 \to \mathbb R_+$.
2. is defined by the equation $u(x) = x_1^px_2^qx_3^{1-p-q}$. 


In the code cell below write this expression for utility and evaluate it for different values of the symbols $p$, $q$ and $x_1$, $x_2, x_3$. 

In [1]:
# Code for example above

p = .2
q = .4
x_1 = 30
x_2 = 30
x_3 = 10

utility = x_1**p * x_2**q * x_3**(1-p-q)
result = f"U({x_1}, {x_2}, {x_3}) = {utility}"
print(result)

U(30, 30, 10) = 19.33182044931763


## Revenue and Cost Functions
Here is a more complicated example of the use of expressions.  Imagine a company produces an item using a 3D printing process.  The output, $y$, of the process are sold to consumers at a price $p$.  Note $y$ is a quantity expressing the number of output units produced. The company's revenue is then given by the equation $r(y) = py$.  

To produce units of $y$ the company must pay a fixed cost for the algorithm, denoted $c_a$, that allows the printer to make a unit and the cost of the printer itself, denoted $c_d$.  In addition each unit of $y$ has an additional variable cost which depends on the amounts of different inputs used in the process. We will let $w$ be the hourly wage rate paid to the operator of the machine, $e$ the cost per kilowatt of energy to run the machine, and $r$ the cost of a foot of filament (a raw material). We can now write the cost of producing output $y$ as with the following equation 
> $c(L, E, R, e, r) = (c_a + c_d) + (wL + eE + rR)$ where $L$ is the labor hours needed to run the machine to produce y units, $E$ is the energy used, and $R$ is the filament used. 

If the prices (w, e, r) are fixed we assume the firm tries to minimize the cost of producing y by choosing a combination (L, E, R) that produce y at minimal cost $c(L, E, R, e, r) = (c_a + c_d) + (wL + eE + rR)$.  We can denote this minimal cost C(y).

We can now define the profit function of the company as 

> $\pi (y) = r(y) - C(y)$. 

In [2]:
# profit equation for example above

p = 50
y = 500
revenue = p*y

c_a = 20
c_d = 3000
cost = c_a + c_d + y**1.5

profit = revenue-cost

result = f"For {y} units produced, profit = {revenue} - {cost} = {profit}"
print(result)

For 500 units produced, profit = 25000 - 14200.339887498949 = 10799.660112501051


>**Note;** Maximum profit is at the point, y*,  where $\pi'(y^*) = r'(y^*) - C'(y^*) = 0$.  Rewriting this equation we get $r'(y^*) = C'(y^*)$.  We note that $r'(y) = p$ and $c'(y) = 1.5y^{1/2}$. 

### Exercise
Solve for the profit maximizing y* in the code cell below.

In [3]:
# p = 1.5y**0.5
# p**2 = (1.5**2)y
y_optimal = int(p**2/(1.5**2))

revenue = p*y_optimal
cost = c_a + c_d + y_optimal**1.5
profit = revenue-cost

result = f"For {y_optimal} units produced, profit = {revenue} - {cost} = {profit}"
print(result)

For 1111 units produced, profit = 55550 - 40051.48162037269 = 15498.518379627312


Are you sure y = 1111 is optimal?  Maybe we didn't take the derivative correctly, or maybe you didn't solve for y_optimal correctly, or maybe we found a minimum since we didn't check the second order condition for a maximum.  Lets do a little sensitivity analysis to check.  To do this try setting y to one unit below the optimum and one unit above the optimum to show that these numbers produce lower profits. 

In [4]:
y = 1110
revenue = p*y
cost = c_a + c_d + y**1.5
profit = revenue-cost

result = f"For {y} units produced, profit = {revenue} - {cost} = {profit}"
print(result)

y = 1112
revenue = p*y
cost = c_a + c_d + y**1.5
profit = revenue-cost

result = f"For {y} units produced, profit = {revenue} - {cost} = {profit}"
print(result)


For 1110 units produced, profit = 55500 - 40001.495372686055 = 15498.504627313945
For 1112 units produced, profit = 55600 - 40101.49036918554 = 15498.509630814457


# Economic Functions in Python

## Utility Functions Revisited
In **section 6.2** we present the concept of a utility function.  Now that we know how to define and use functions lets build the Cobb-Douglas utility function in Python.
>Utility
>1. $u:$$\mathbb R_+^2 \to \mathbb R_+$.
2. is defined by the equation $u(x) = x_1^q x_2^{1-q}$,<br> where 0 < q <1. 

In the code cell below we have started writing the Cobb-Douglass utility function.  Your job is to complete the function.  Notice we have added a comment called `#TODO` which indicates what needs to be done. 

In [6]:
def cobb_douglass_u(q, x_1, x_2):
    """Calculates utility = x_1**q * x_2**(1-q)
    
        args: 
        
        q, float, coeeficient between 0 and 1
        x_1, int, quantity of good 1
        x_2, int, quantity of good 2
          
    returns: utility, float
    
    """
    #TODO - calculate utility and return it 
    return 0.0

Once you are done run the test code below.  

In [7]:
# This should return 5.0 when you have finished the #TODO above.

cobb_douglass_u(.25, 5, 5)

0.0